### ISSUE TITLE
Multiline autocomplete that supports multiple partial text matches

### ISSUE THREAD

**Issue by dkario** [adobe/react-spectrum#5216]
### Provide a general summary of the feature here

Requesting support for a feature similar to @-mentioning in Slack, Outlook, etc, a multiline textbox that can trigger an associated popover + listbox like ComboBox, either by pressing a trigger button or typing a reserved, configurable character like `@` or `/`.

### 🤔 Expected Behavior?

**Character trigger**
* Typing the character trigger opens the listbox, if it's placed at the beginning of the input or preceded by whitespace or configurable special characters. For example, "Hi, @" opens it, but "Hi@" does not.
* Clicking the trigger button inserts the trigger character into the input at the cursor, or by default at the end of the input.
* Unlike useComboBox, probably do not support menuTrigger of "input" or "focus".

**Partial text match**
* When the listbox is open, as the user types, collection items are filtered by the characters immediately following the trigger character. For example, "What do you think, @{user-input}" will only filter on "{user-input}".
* Support or allow configuration for spaces and punctuation in {user-input}, like [Lexical](https://github.com/facebook/lexical/blob/9a679ecb2fa3ca999772e2a170f4ad15d19d4635/packages/lexical-playground/src/plugins/MentionsPlugin/index.tsx).

**Replace partial input**
* When a listbox option is selected, the option's text replaces the string "@{user-input}" including the trigger character, and the selected key is immediately cleared. All text before the trigger character and after {user-input} remains the same.

**Keyboard + aria**
* To support multiline input, when the listbox is closed, up and down arrow no longer open the listbox. Alt+up and alt+down can open it instead, similar to how we open the Calendar from a date segment in DatePicker.
* Text field should have `role="textbox"` and not `role="combobox",` but should treat `aria-autocomplete` and `aria-activedescendant` the same as ComboBox.

**Style**
* Trigger button visual placement is TBD. Options include: above/below text input and alone, above/below text input and in a configurable toolbar, overlaid on text input.
* The listbox is as wide as the text input, regardless of where the trigger button lives.
* Explore support for different visible treatment and interactivity of autocompleted text, like an @-mentioned person in a Slack message that's still being typed.

### 😯 Current Behavior

The closest alternative available today is ComboBox.

* No support for partial text matching. The entire input is used for filtering.
* When a listbox item is selected, the entire text input is replaced, and the selectedKey is set or controlled.
* Trigger button has fixed location and ChevronDown icon
* `role="combobox"`

### 💁 Possible Solution

Refactor useComboBox and useComboBoxState, to extract some core logic that can also apply to potential new hooks.
* Separate listbox and trigger management from text filtering logic.
* Create new hooks, component, and/or RAC.

### 🔦 Context

I prototyped this for an internal Adobe project.

### 🧢 Your Company/Team

Adobe

**Comment by snowystinger** [comment]
Are you not able to do the filtering you want using https://react-spectrum.adobe.com/react-spectrum/ComboBox.html#custom-filtering

**Comment by dkario** [comment]
@snowystinger yup I'm doing something similar to that for the filtering, but when an item's selected, I still need only part of the input to get replaced

**Comment by dkario** [comment]
Here's a list of things I've modified so far to get an MVP working. I got to a stage where it no longer makes sense to keep track of all this inside code comments in my files, and I'm not ready to open an RSP PR yet, so I'll continue documenting this as I go to make the eventual transition to RSP easier. Not an amazing solution but works for now.

## ComboBox.tsx
Change | Remove/Add
-- | --
Removed ChevronDownMedium icon | -
Removed All mobile functionality: useIsMobileDevice, MobileComboBox, for now | -
Removed Field, FieldButton | -
Removed FocusRing, no longer propagate autoFocus prop to FocusRing | -
Removed Its child InputGroup div | -
Removed useHover props that were spread on that div | -
Removed ref that was placed on that div | -
Import new local useComboBox and useComboBoxState | -
Replace comboboxStyles, styles, and textFieldStyles with empty objects (for now) | -
AtSignMentionsRegex and all helper variables taken from Lexical for now | +
parseMention custom logic | +
ComboBoxBase menuTrigger prop defaults to ‘manual’ | +
popoverRef and buttonRef default to null | +
Add controlled state for selectedKey | +
inputValue is controlled by parent ChatInput | +
domRef no longer placed on Field, since Field is removed. only used in new custom call to useImperativeHandle that copies TextFieldBase by exposing getInputElement | +
handleSelectionChange does partial text replacement based on parseMention output and calls setSelectedKey | +
useEffect resets selectedKey to null whenever it changes | +
Call useComboBoxState with new overrides to …props: menuTrigger, selectedKey, onSelectionChange: handleSelectionChange | +
handleKeyDown passed to useComboBoxState: suppress onKeyDown when hitting enter while menu is open, since that will trigger an Assistant query. this is business logic for the rail | +
menu width + resizing: align to inputWidth since we removed input group button | -
useResizeObserver uses inputRef, not domRef (deleted) | +
remove classnames from Popover | -
Popover triggerRef is now inputRef to align itself | +
removed placeholder warning since we have to use one :( | -
Spread …otherProps onto TextFieldBase | +
Add multiLine prop to TextFieldBase | +
Add autoComplete=“off” to TextFieldBase | +
Add ref={ref} to TextFieldBase, the ref that used to be on Field parent for resizing | +
Remove 'spectrum-InputGroup-field' from TextFieldBase and 'spectrum-InputGroup-input' from inputClassNames prop, 'spectrum-FieldButton' from ActionButton | -
Remove disableFocusRing from TextFieldBase | -
Absolute position for trigger ActionButton | +

## useComboBox.ts
Change | Remove/Add
-- | --
Remove default completionMode = ‘suggest' | -
Replaced RSP’s string formatter with Unified Shell and react-intl | +
In onKeyDown, remove handlers for up and down arrows opening the menu, since they’re used to navigate multiline input. Decided to delete earlier guidance to have alt+up and alt+down open menu, since it no longer makes sense: opening the menu needs to add an @ character if one doesn’t exist, and that’s unintuitive for alt+arrow. Plus, future menus can be dynamic, not just this @-mention one, each triggered by diff keys or buttons. Not just a simple autocomplete/combobox with one single type of menu. | -
In onPress, add character trigger to input if not there already, then move cursor to after the character | +
Remove excludeFromTabOrder from buttonProps | -
Change role from ‘combobox’ to ‘textbook’ in inputProps | +

## useComboBoxState.ts
Change | Remove/Add
-- | --
In the useEffect that runs on every render, change open/close condition. Keep first boolean checks up thru inputValue !== lastValue.current. Then, if menu closed and a mention is parsed, open it. If menu is open and a mention is no longer parsed, close it. | -

**Comment by devongovett** [comment]
Gonna close this for now. It ended up being built internally.
